botorch_community/models/vbll_helper.py

```diff
@@ -0,0 +1,487 @@
+    def __init__(self, loc, chol):
+        super(Normal, self).__init__(loc, chol)
+
+    @property
+    def mean(self):
+        return self.loc
+
+    @property
+    def var(self):
+        return self.scale**2
+
+    @property
+    def chol_covariance(self):
+        return torch.diag_embed(self.scale)
+
+    @property
+    def covariance_diagonal(self):
+        return self.var
+
+    @property
+    def covariance(self):
+        return torch.diag_embed(self.var)
+
+    @property
+    def precision(self):
+        return torch.diag_embed(1.0 / self.var)
+
+    @property
+    def logdet_covariance(self):
+        return 2 * torch.log(self.scale).sum(-1)
+
+    @property
+    def logdet_precision(self):
+        return -2 * torch.log(self.scale).sum(-1)
+
+    @property
+    def trace_covariance(self):
+        return self.var.sum(-1)
+
+    @property
+    def trace_precision(self):
+        return (1.0 / self.var).sum(-1)
+
+    def covariance_weighted_inner_prod(self, b, reduce_dim=True):
+        assert b.shape[-1] == 1
+        prod = (self.var.unsqueeze(-1) * (b**2)).sum(-2)
+        return prod.squeeze(-1) if reduce_dim else prod
+
+    def precision_weighted_inner_prod(self, b, reduce_dim=True):
+        assert b.shape[-1] == 1
+        prod = ((b**2) / self.var.unsqueeze(-1)).sum(-2)
+        return prod.squeeze(-1) if reduce_dim else prod
+
+    def __add__(self, inp):
+        if isinstance(inp, Normal):
+            new_cov = self.var + inp.var
+            return Normal(
+                self.mean + inp.mean, torch.sqrt(torch.clip(new_cov, min=1e-12))
+            )
+        elif isinstance(inp, torch.Tensor):
+            return Normal(self.mean + inp, self.scale)
+        else:
+            raise NotImplementedError(
+                "Distribution addition only implemented for diag covs"
+            )
+
+    def __matmul__(self, inp):
+        assert inp.shape[-2] == self.loc.shape[-1]
+        assert inp.shape[-1] == 1
+        new_cov = self.covariance_weighted_inner_prod(
+            inp.unsqueeze(-3), reduce_dim=False
+        )
+        return Normal(self.loc @ inp, torch.sqrt(torch.clip(new_cov, min=1e-12)))
+
+    def squeeze(self, idx):
+        return Normal(self.loc.squeeze(idx), self.scale.squeeze(idx))
+
+
+class DenseNormal(torch.distributions.MultivariateNormal):
+    def __init__(self, loc, cholesky):
+        super(DenseNormal, self).__init__(loc, scale_tril=cholesky)
+
+    @property
+    def mean(self):
+        return self.loc
+
+    @property
+    def chol_covariance(self):
+        return self.scale_tril
+
+    @property
+    def covariance(self):
+        return self.scale_tril @ tp(self.scale_tril)
+
+    @property
+    def inverse_covariance(self):
+        warnings.warn(
+            "Direct matrix inverse for dense covariances is O(N^3), consider using eg inverse weighted inner product"
+        )
+        return tp(torch.linalg.inv(self.scale_tril)) @ torch.linalg.inv(self.scale_tril)
```

Comment:
Ugh, this is pretty bad. You're calling `inv` on the same tensor twice. Also, you already have the Cholesky decomposition here, why are you calling `inv` in the first place? You can do this with a simple triangular solve.
```suggestion
        Eye = torch.eye(self.scale_tril.shape[-1], device=self.scale_tril.device, dtype=self.scale_tril.dtype)
        W = torch.linalg.solve_triangular(self.scale_tril, Eye, upper=False)
        return tp(W) @ W
```

Reply:
uff, thank you! Will also push this to the vbll repo